diagnostics/include/performance_handler.h

```diff
@@ -253,39 +240,88 @@ struct Histogram {
   std::chrono::system_clock::time_point snapshot_end;
 
   // The histogram interval starting from when the last snapshot was taken
-  hdr_histogram* snapshot = nullptr;
-
+  histogram<axes_t> snapshot;
   // History doesn't include the latest snapshot.
-  hdr_histogram* history = nullptr;
+  histogram<axes_t> history;
+  std::mutex mutex_;
 };
 
 using Name = std::string;
 using Histograms = std::map<Name, Histogram>;
 
-// Useful data extracted from hdr_histogram(s)
+// Useful data extracted from histogram(s)
 //
 // By extracting this data we can return it in a thread-safe manner, without having to copy the
 // entire histogram.
 //
 // We specifically don't return the average, as it's a completely meaningless metric.
 struct HistogramValues {
-  HistogramValues(hdr_histogram* h) : memory_used(hdr_get_memory_size(h)) {
-    max = hdr_max(h);
+  HistogramValues(histogram<axes_t> h) : memory_used(sizeof(h)) {
+    max = get_max_value(h);
     if (max == 0) return;  // Histogram is empty
-    min = hdr_min(h);
-    count = h->total_count;
-    pct_10 = hdr_value_at_percentile(h, 10);
-    pct_25 = hdr_value_at_percentile(h, 25);
-    pct_50 = hdr_value_at_percentile(h, 50);
-    pct_75 = hdr_value_at_percentile(h, 75);
-    pct_90 = hdr_value_at_percentile(h, 90);
-    pct_95 = hdr_value_at_percentile(h, 95);
-    pct_99 = hdr_value_at_percentile(h, 99);
-    pct_99_9 = hdr_value_at_percentile(h, 99.9);
-    pct_99_99 = hdr_value_at_percentile(h, 99.99);
-    pct_99_999 = hdr_value_at_percentile(h, 99.999);
-    pct_99_9999 = hdr_value_at_percentile(h, 99.9999);
-    pct_99_99999 = hdr_value_at_percentile(h, 99.99999);
+    min = get_min_value(h);
+    count = get_total_element(h);
+    pct_10 = get_bin_value_greater_than_percent(h, 10);
+    pct_25 = get_bin_value_greater_than_percent(h, 25);
+    pct_50 = get_bin_value_greater_than_percent(h, 50);
+    pct_75 = get_bin_value_greater_than_percent(h, 75);
+    pct_90 = get_bin_value_greater_than_percent(h, 90);
+    pct_95 = get_bin_value_greater_than_percent(h, 95);
+    pct_99 = get_bin_value_greater_than_percent(h, 99);
+    pct_99_9 = get_bin_value_greater_than_percent(h, 99.9);
+    pct_99_99 = get_bin_value_greater_than_percent(h, 99.99);
+    pct_99_999 = get_bin_value_greater_than_percent(h, 99.999);
+    pct_99_9999 = get_bin_value_greater_than_percent(h, 99.9999);
+    pct_99_99999 = get_bin_value_greater_than_percent(h, 99.99999);
+  }
+
+  int64_t get_total_element(const histogram<axes_t>& h) {
+    // Compute the cumulative sum of bin heights
+    int64_t total_elem = 0;
+    int count = h.axis().size();
+    for (auto i = 0; i < count; ++i) {
+      total_elem += h.at(i);
+    }
+    return total_elem;
+  }
+
+  int64_t get_max_value(const histogram<axes_t>& h) {
+    int last_bin = 0;
+    for (auto i = h.axis().size() - 1; i >= 0; i--) {
+      if (h.at(i) > 0) {
+        last_bin = i;
+        break;
+      }
+    }
+    return last_bin;
```

Comment:
Once you find max bin, Is there a way to get exact max value from within bin? I think it should be there.